fix: backwards compatibility for pipeline spec task_name

zazulam opened the pull request:
**Description of your changes:**
This change is to add a check on whether the API server version contains the necessary backend changes to process the `task_name` values on the TaskSpec in the Pipeline IR. Since the `kfp-pipeline-spec` and the `kfp` packages are more tied together I figured the best way to detect the version of the backend was to just check the `kfp-server-api` package version since that is tied the backend version that would handle the new field logic. Please let me know if you think there's a better way to handle this for now. I think there should be a discussion around the approach for handling backwards compatibility, not sure if it should be tied in with the testing/CI discussions.

For reference: https://github.com/kubeflow/pipelines/blob/master/sdk/python/requirements.in#L13-L19


**Checklist:**
- [x] You have [signed off your commits](https://www.kubeflow.org/docs/about/contributing/#sign-off-your-commits)
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 
<!--
   PR titles examples:
    * `fix(frontend): fixes empty page. Fixes #1234`
       Use `fix` to indicate that this PR fixes a bug.
    * `feat(backend): configurable service account. Fixes #1234, fixes #1235`
       Use `feat` to indicate that this PR adds a new feature. 
    * `chore: set up changelog generation tools`
       Use `chore` to indicate that this PR makes some changes that users don't need to know.
    * `test: fix CI failure. Part of #1234`
        Use `part of` to indicate that a PR is working on an issue, but shouldn't close the issue when merged.
-->

droctothorpe commented:
Thanks for tackling this so quickly, @zazulam!

I think it's a good idea to do some manual validation from local for the following two scenarios:

1. Updated SDK, previous version of API
2. Updated API, previous version of SDK

We can look into automating that kind of backwards compatibility validation in a discrete PR.

zazulam commented:
> Thanks for tackling this so quickly, @zazulam!
> 
> I think it's a good idea to do some manual validation from local for the following two scenarios:
> 
> 1. Updated SDK, previous version of API
> 2. Updated API, previous version of SDK
> 
> We can look into automating that kind of backwards compatibility validation in a discrete PR.

@droctothorpe, yea I had modified my deployment to have the api server + the driver & launcher point to 2.5.0, the local version of the `kfp-server-api` is still set to 2.5.0 so it was able to launch the pipelines from the v2 samples set where only the ones that failed were relying on what is on master for the backend, like the fixes for collected and display_name.

zazulam commented:
After discussing offline with others, changed course and opted to remove the new field from the pipeline spec and the logic in the sdk to add it in and place it in the compiler to hand off to the driver. This way the amount of code necessary for the original unique task name is reduced and sdk backwards compatibility is preserved for 2.13 when the new release will be cut (if this gets merged in by then).

This meant having to recompile all the test cases that were added before to remove the field, hence all the yaml diffs 😅 

It seems to me that the backend test failing `TestCreatePipelineV1_LargeFile` with the error below means there _might_ be some mismatch of a stub that is being tested against the pipeline server.
```
=== RUN   TestCreatePipelineV1_LargeFile
    pipeline_server_test.go:104: 
        	Error Trace:	/home/runner/work/pipelines/pipelines/backend/src/apiserver/server/pipeline_server_test.go:104
        	Error:      	Expected nil, but got: &util.UserError{internalError:(*errors.withStack)(0xc0014ba5b8), externalMessage:"unknown template format", externalStatusCode:0x3}
        	Test:       	TestCreatePipelineV1_LargeFile
    pipeline_server_test.go:105: 
        	Error Trace:	/home/runner/work/pipelines/pipelines/backend/src/apiserver/server/pipeline_server_test.go:105
        	Error:      	Expected value not to be nil.
        	Test:       	TestCreatePipelineV1_LargeFile
--- FAIL: TestCreatePipelineV1_LargeFile (0.17s)
```

zazulam commented:
Turns out we test against a pipeline IR from master:  
[pipeline](https://github.com/kubeflow/pipelines/blob/master/sdk/python/test_data/pipelines/xgboost_sample_pipeline.yaml#L762-L764) in question from master with TaskName still set.

[test](https://github.com/kubeflow/pipelines/blob/master/backend/src/apiserver/server/pipeline_server_test.go#L87-L102) using said pipeline.

Not sure how to proceed, the fix works it just fails that test.

cc: @droctothorpe @mprahl @HumairAK

zazulam commented:
Pulled in the changes from master and now the tests are failing from kind not launching & protobuf dep issues. I can't figure out the diff that would be causing the kind issue or the protobuf problem, honestly feels like the CI is just working against me 🙃 

<img width="823" height="310" alt="image" src="https://github.com/user-attachments/assets/df3820e0-ed34-49a2-843a-1976a8d1d8b6" />
<img width="632" height="409" alt="image" src="https://github.com/user-attachments/assets/3ee144e6-9588-4836-a468-e1e64a4bdef5" />

zazulam commented:
@droctothorpe @mprahl managed to get the CI passing 🙏

droctothorpe commented:
Kudos, Mike! This LGTM. Thanks for resolving the problematic test! 

/approve
